File: 0199_binary_tree_right_side_view/right_side_view.rb
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val = 0, left = nil, right = nil)
        @val = val
        @left = left
        @right = right
    end
end

def right_side_view(root)
    # Time: O(n)
    # Space: O(n)
    return [] if root.nil?
    q, res = [], []
    q.append(root)
    until q.empty? do
        size, val = q.length, 0
        size.times do
          popped = q.shift
          val = popped.val
          q.append(popped.left) unless popped.left.nil?
          q.append(popped.right) unless popped.right.nil?
        end
        res.append(val)
    end
    res
end

File: 0199_binary_tree_right_side_view/right_side_view.py
from collections import deque
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def rightSideView(root: TreeNode) -> List[int]:
    """
    Time: O(n)
    Space: O(n)
    """
    if not root:
        return []
    q = deque([root])
    res = []
    while q:
        size, val = len(q), 0
        for _ in range(size):
            popped = q.popleft()
            val = popped.val
            if popped.left:
                q.append(popped.left)
            if popped.right:
                q.append(popped.right)
        res.append(val)
    return res
        